File: csexts/dumppo.cs
using System;
using System.Collections.Generic;
using System.Text;
using DllExporter;
using System.Runtime.InteropServices;

namespace CsExts
{
    public static partial class CsExts
    {
        [DllExport]
        public static int dumppo(IntPtr debugClient4, IntPtr args)
        {
            const int DEBUG_VALUE_INT32 = 3;
            UnsafeNativeMethods.DEBUG_VALUE value;
            int cb;
            uint remainder;
            int bufSize = Marshal.SizeOf(typeof(ulong));
            IntPtr buffer = Marshal.AllocHGlobal(bufSize);

            UnsafeNativeMethods.IDebugClient4 client = (UnsafeNativeMethods.IDebugClient4)Marshal.GetTypedObjectForIUnknown(debugClient4, typeof(UnsafeNativeMethods.IDebugClient4));
            string argument = Marshal.PtrToStringAnsi(args);
            UnsafeNativeMethods.IDebugControl control = (UnsafeNativeMethods.IDebugControl)client;
            UnsafeNativeMethods.IDebugDataSpaces dataSpaces = (UnsafeNativeMethods.IDebugDataSpaces)client;

            control.Evaluate(argument, DEBUG_VALUE_INT32, out value, out remainder);
            dataSpaces.ReadVirtual((ulong)value.I32, buffer, bufSize, out cb);
            Console.WriteLine("Example4, PurchaseOrder: Id={0}", Marshal.ReadInt32(buffer));

            Marshal.FreeHGlobal(buffer);
            dataSpaces = null;
            client = null;
            control = null;
            return 0;
        }
    }
}

File: hwexts/dumppo.cpp
#include "hwexts.h"
#include "PurchaseOrder.h"
#include "out.hpp"

//HRESULT CALLBACK dumppo(PDEBUG_CLIENT4 Client, PCSTR args)
//{
//    ULONG cb;
//    ULONG64 Address;
//    ULONG   Buffer[1];
//
//    Address = GetExpression(args);
//    ReadMemory((ULONG_PTR)Address, &Buffer, sizeof(Buffer), &cb);
//    dprintf("Example1, PurchaseOrder: Id=%d\n", Buffer[0]);
//
//    return S_OK;
//}

//HRESULT CALLBACK dumppo(PDEBUG_CLIENT4 Client, PCSTR args)
//{
//    ULONG cb;
//    ULONG64 Address;
//    PurchaseOrder po;
//
//    // ReadMemory to struct
//    Address = GetExpression(args);
//    ReadMemory((ULONG_PTR)Address, &po, sizeof(po), &cb);
//    dprintf("Example2, PurchaseOrder: Id=%d\n", po.Id);
//
//    return S_OK;
//}

HRESULT CALLBACK dumppo(PDEBUG_CLIENT4 Client, PCSTR args)
{
    ULONG cb;
    PurchaseOrder po;
    PDEBUG_DATA_SPACES DataSpaces;
    PDEBUG_CONTROL Control; 
    DEBUG_VALUE DebugValue; 
    ULONG Remainder;

    Client->QueryInterface(__uuidof(IDebugDataSpaces), (void **)&DataSpaces);
    Client->QueryInterface(__uuidof(IDebugControl), (void **)&Control);

    Control->Evaluate(args, DEBUG_VALUE_INT32, &DebugValue, &Remainder);
    DataSpaces->ReadVirtual(DebugValue.I32, &po, sizeof(po), &cb);
    dprintf("Example3, PurchaseOrder: Id=%d\n", po.Id);

    Control->Release();
    DataSpaces->Release();

    return S_OK;
}
